tools/testsys/src/vmware_k8s.rs

```diff
@@ -0,0 +1,242 @@
+};
+use crate::error::{self, Result};
+use crate::sonobuoy::sonobuoy_crd;
+use bottlerocket_types::agent_config::{
+    CreationPolicy, K8sVersion, VSphereK8sClusterConfig, VSphereK8sClusterInfo, VSphereVmConfig,
+};
+use maplit::btreemap;
+use model::{Crd, DestructionPolicy};
+use pubsys_config::vmware::Datacenter;
+use snafu::OptionExt;
+use std::collections::BTreeMap;
+use std::str::FromStr;
+
+/// A `CrdCreator` responsible for creating crd related to `vmware-k8s` variants.
+pub(crate) struct VmwareK8sCreator {
+    pub(crate) region: String,
+    pub(crate) datacenter: Datacenter,
+    pub(crate) ova_name: String,
+    pub(crate) encoded_mgmt_cluster_kubeconfig: String,
+}
+
+#[async_trait::async_trait]
+impl CrdCreator for VmwareK8sCreator {
+    /// Use the provided OVA name for the image id.
+    fn image_id(&self, _: &CrdInput) -> Result<String> {
+        Ok(self.ova_name.to_string())
+    }
+
+    /// Use standard naming conventions to predict the starting OVA.
+    async fn starting_image_id(&self, crd_input: &CrdInput) -> Result<String> {
+        Ok(format!(
+            "bottlerocket-{}-{}-{}",
+            crd_input.variant,
+            crd_input.arch,
+            crd_input
+                .starting_version
+                .as_ref()
+                .context(error::InvalidSnafu {
+                    what: "The starting version must be provided for migration testing"
+                })?
+        ))
+    }
+
+    /// Create an vSphere K8s cluster CRD with the `cluster_name` in `cluster_input`.
+    async fn cluster_crd<'a>(&self, cluster_input: ClusterInput<'a>) -> Result<CreateCrdOutput> {
+        let control_plane_endpoint = cluster_input
+            .crd_input
+            .config
+            .control_plane_endpoint
+            .as_ref()
+            .context(error::InvalidSnafu {
+                what: "The control plane endpoint is required for vmware cluster creation.",
+            })?;
+        let labels = cluster_input.crd_input.labels(btreemap! {
+            "testsys/type".to_string() => "cluster".to_string(),
+            "testsys/cluster".to_string() => cluster_input.cluster_name.to_string(),
+            "testsys/controlPlaneEndpoint".to_string() => control_plane_endpoint.to_string(),
+        });
+
+        // Check if the cluster already has a CRD
+        if let Some(cluster_crd) = cluster_input
+            .crd_input
+            .existing_crds(
+                &labels,
+                &[
+                    "testsys/cluster",
+                    "testsys/type",
+                    "testsys/controlPlaneEndpoint",
+                ],
+            )
+            .await?
+            .pop()
+        {
+            return Ok(CreateCrdOutput::ExistingCrd(cluster_crd));
+        }
+
+        // Check if an existing cluster is using this endpoint
+        let existing_clusters = cluster_input
+            .crd_input
+            .existing_crds(&labels, &["testsys/type", "testsys/controlPlaneEndpoint"])
```

Comment:
Don't you have to pass the same list you passed in the line above? Also is this checking or retrieving existing clusters? It seems like the latter since there isn't any "check" like in the line above.

Reply:
This one is just determining if any clusters are using the same ip as this one. Only 1 cluster can use an ip at a time, so this cluster will wait until the conflicting cluster is cleaned up before being created.

Reply:
> until the conflicting cluster is cleaned up

Does this mean the user has to clean up the cluster? If so, do we provide a mechanism obvious enough to the users (`testsys logs` or what have you) so that they can figure out the cluster isn't unique? or will testsys hang with an enigmatic error?

Reply:
The clusters are automatically cleaned up once all tests that use them have passed.